### PR TITLE
[Refactor][Ops] Refactor MoE mlp and quantization methods

### PR DESCRIPTION
**Pull request by earthmanylf**
### What this PR does / why we need it?

Come from #13318 and #13220
Right now, moe_mlp has too many branches about different quantization method and different activations. In this PR, we split moe_mlp's methods into differnet quantization method. 

### Does this PR introduce _any_ user-facing change?

No. This is an internal refactor of the fused MoE runtime contract and its test coverage.

### How was this patch tested?
The following models are tested and can be curled successfully:
A3:
float: qwen30b-a3b
w8a8: qwen30b-a3b-w8a8
w4a8: kimi2.6-w4a8
megamoe, dynamic eplb: DeepSeek-V4-Flash-w8a8-mtp
A5:
w8a8mxfp: step3.7-w8a8mxfp
w4a8mxfp: qwen3-30b-w4a8mxfp
w4a4mxfp: glm5.1-w4a4mxfp

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/ops/fused_moe/moe_utils.py [hunk 1]
```diff
@@ -178,3 +180,85 @@ def get_moe_num_logical_experts(
...
+def _custom_gmm_swiglu_enabled(fusion, dynamic_eplb, activation=None):
+    return (
+        fusion
+        and dynamic_eplb
+        and getattr(activation, "value", activation) != "swigluoai_uninterleave"
+        and enable_custom_op()
+    )
+
+
+def _gmm_swiglu_quant_fusion_enabled(use_mxfp_quant, fusion, dynamic_eplb, activation=None):
```
Comment (zhangc10) [inline vllm_ascend/ops/fused_moe/moe_utils.py]: This function can be splitted. Only fusion and dynamic_eplb is useful. use_mxfp_quant and activation is useless after the refactoring.

File: vllm_ascend/ops/fused_moe/moe_utils.py [hunk 2]
```diff
@@ -178,3 +180,85 @@ def get_moe_num_logical_experts(
         return int(num_logical_experts)
 
     return int(num_experts - global_redundant_expert_num - num_shared_experts)
+
+
+def _custom_gmm_swiglu_enabled(fusion, dynamic_eplb, activation=None):
```
Comment (zhangc10) [inline vllm_ascend/ops/fused_moe/moe_utils.py:166]: This function can be splitted. Activation is useless after the refactoring.

File: vllm_ascend/ops/fused_moe/routed_experts.py [hunk 3]
```diff
@@ -168,17 +159,145 @@ def apply(
...
                 swiglu_alpha=layer.swiglu_alpha,
                 swiglu_beta=layer.swiglu_beta,
                 lora_context=getattr(layer, "_ascend_moe_lora_context", None),
+            ),
+            quant_method=self,
+        )
+
+    # ---- MLP gmm hooks (activation orchestrated by MoeActionMethod) ----
+
+    def get_moe_weights(self, layer):
```
Comment (zhangc10) [inline vllm_ascend/ops/fused_moe/routed_experts.py]: This function seems not consistent with the original code in `AscendUnquantizedFusedMoEMethod`'s `apply` method

File: vllm_ascend/ops/fused_moe/routed_experts.py [hunk 4]
```diff
@@ -168,17 +159,145 @@ def apply(
...
+            w1=w1,
+            w2=w2,
+            w1_scale=[torch.tensor([], dtype=torch.int64)],
+            w2_scale=[torch.tensor([], dtype=torch.int64)],
+            w1_scale_bias=[torch.tensor([], dtype=torch.float32)],
+            w2_scale_bias=[torch.tensor([], dtype=torch.float32)],
         )
-        return final_hidden_states
+
+    def _get_mlp_weights(self, layer):
```
Comment (zhangc10) [inline vllm_ascend/ops/fused_moe/routed_experts.py]: `_get_mlp_weights` processes both `w1` and `w2`, but only one of them is used. Either in `apply_gmm1` or `apply_gmm2`. Why not split it into two functions, one for `w1`, another for `w2`.

File: vllm_ascend/ops/fused_moe/routed_experts.py [hunk 5]
```diff
@@ -168,17 +159,145 @@ def apply(
...
+
+    def apply_gmm1(self, mlp_compute_input: MoEMlpComputeInput):
+        layer = mlp_compute_input.layer
+        w1, _ = self._get_mlp_weights(layer)
+        w1 = self._maybe_transpose(w1, mlp_compute_input.need_trans)
+        hidden_states = mlp_compute_input.hidden_states
+        w1_bias = layer.w13_bias if self.moe.has_bias else None
+        gate_up_out = torch_npu.npu_grouped_matmul(
+            x=[hidden_states],
+            weight=w1 if isinstance(w1, list) else [w1],
```
Comment (zhangc10) [inline vllm_ascend/ops/fused_moe/routed_experts.py:211]: In the original `unquant_apply_mlp`, w1 is not a list. Maybe it's redundant to add the `isinstance(w1, list)`

File: vllm_ascend/quantization/methods/base.py [hunk 6]
```diff
@@ -285,3 +261,82 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+            # A16 quantization doesn't need to quant hidden_states
+            return hidden_states, None
+        elif dynamic_scale is None:
+            # When hidden_states haven't been quanted, we need to quant hidden_states.
+            act_quant_type = torch.int8 if self.quant_type in [QuantType.W8A8, QuantType.W4A8] else torch.float8_e4m3fn
+            use_mxfp_quant = self.quant_type in (
+                QuantType.W8A8MXFP,
+                QuantType.W4A4MXFP,
+                QuantType.W4A8MXFP,
+                QuantType.W4A16MXFP,
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/base.py]: `W4A16MXFP` is not needed here.

File: vllm_ascend/quantization/methods/w4a16.py [hunk 7]
```diff
@@ -318,3 +319,51 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+            w2=layer.w2_weight_packed,
+            w1_scale=layer.w13_weight_scale,
+            w2_scale=layer.w2_weight_scale,
+            w1_offset=layer.w13_weight_offset,
+            w2_offset=layer.w2_weight_offset,
+        )
+
+    def apply_gmm1(self, mlp_compute_input: MoEMlpComputeInput):
+        # A16: activations stay unquantized; gmm1 dequantizes with antiquant.
+        hidden_states, _ = self._quant_hidden_states(mlp_compute_input.hidden_states, mlp_compute_input.dynamic_scale)
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a16.py]: For specific quant types (e.g., here W4A16), it's better not to use the general `_quant_hidden_states` method. As you know for sure, `hidden_states` stay unquantized.

File: vllm_ascend/quantization/methods/w4a16_mxfp4.py [hunk 8]
```diff
@@ -53,6 +55,8 @@ class AscendW4A16MXFP4FusedMoEMethod(AscendMoEScheme):
 
     supports_eplb = False
     quant_type: QuantType = QuantType.W4A16MXFP
+    act_quant_type: torch.dtype | None = None
+    fused_activations = frozenset({"silu"})
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a16_mxfp4.py]: Consider an empty set for W4A16_MXFP4 (i.e., `fused_activations = frozenset()`). It has a similar logic as W4A16, the the following `apply_gmm1_act_quant` is not needed.

File: vllm_ascend/quantization/methods/w4a8.py [hunk 9]
```diff
@@ -752,6 +728,218 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+            if group_list_type == 0:
+                group_list = torch.cat([group_list[:1], torch.diff(group_list, dim=0)])
+                group_list_type = 1
+            output_dtype = torch.bfloat16
+        else:
+            _, _, _, w2_scale = self._get_mlp_weights(mlp_compute_input.layer)
+            output_dtype = w2_scale[0].dtype if isinstance(w2_scale, list) else w2_scale.dtype
+        return group_list, group_list_type, output_dtype
+
+    def get_moe_weights(self, layer):
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8.py]: It's better to use a more clear method name. `get_moe_weights` is a little confused. B.T.W, the final `return` branch seems for the case of non-fused mc2 (may be redundant)

File: vllm_ascend/quantization/methods/w4a8.py [hunk 10]
```diff
@@ -752,6 +728,218 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+            mlp_compute_input.hidden_states, mlp_compute_input.dynamic_scale
+        )
+        layer = mlp_compute_input.layer
+        w1, w1_scale, _, _ = self._get_mlp_weights(layer)
+        bias1, _ = self._get_mlp_scale_bias(layer)
+        group_list, group_list_type, _ = self._maybe_convert_group_list(mlp_compute_input, bias1)
+        activation = mlp_compute_input.activation
+        act_name = getattr(activation, "value", activation)
+
+        if self.is_per_channel_weight and enable_custom_op() and act_name != "swigluoai_uninterleave":
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8.py]: It seems that `act_name != "swigluoai_uninterleave"` is not needed here.

File: vllm_ascend/quantization/methods/w4a8.py [hunk 11]
```diff
@@ -752,6 +728,218 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+                weight=w1,
+                weight_scale=w1_scale if isinstance(w1_scale, list) else [w1_scale],
+                x_scale=pertoken_scale,
+                group_list=group_list,
+                weight_assist_matrix=bias1,
+                dequant_mode=0,
+                group_list_type=group_list_type,
+                swiglu_limit=mlp_compute_input.swiglu_limit,
+            )
+        elif _custom_gmm_swiglu_enabled(mlp_compute_input.fusion, mlp_compute_input.dynamic_eplb, activation):
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8.py]: This branch is not needed for W4A8. In the original code, when `use_w4a8_per_channel_gmm_swiglu` is True, it only enter the first branch, which calls `torch.ops._C_ascend.grouped_matmul_swiglu_quant_v2`.

File: vllm_ascend/quantization/methods/w4a8.py [hunk 12]
```diff
@@ -752,6 +728,218 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+                weight=w1,
+                weight_scale=w1_scale,
+                x_scale=pertoken_scale,
+                group_list=cumsum_group_list(group_list, group_list_type, 0),
+                bias=bias1,
+                swiglu_limit=mlp_compute_input.swiglu_limit,
+            )
+        elif _gmm_swiglu_quant_fusion_enabled(
+            False, mlp_compute_input.fusion, mlp_compute_input.dynamic_eplb, activation
+        ):
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8.py]: Same as above, this branch is not needed for W4A8

File: vllm_ascend/quantization/methods/w4a8.py [hunk 13]
```diff
@@ -752,6 +728,218 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+            # Unified fallback: soft-quant gmm1 + swiglu_quant.
+            hidden_states = self._soft_gmm1(mlp_compute_input, hidden_states, pertoken_scale)
+            hidden_states, swiglu_out_scale = self.apply_swiglu_quant(mlp_compute_input, hidden_states)
+
+        # The input dispatch buffer is either consumed (pre-quantized reuse) or
+        # replaced by a fresh quantized copy; release it like the legacy path.
+        dispose_tensor(mlp_compute_input.hidden_states)
+        return hidden_states, swiglu_out_scale
+
+    def _soft_gmm1(self, mlp_compute_input: MoEMlpComputeInput, hidden_states: torch.Tensor, pertoken_scale):
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8.py]: The method name `_soft_gmm1` seems confused. What does soft mean?

File: vllm_ascend/quantization/methods/w4a8/w4a8.py [hunk 14]
```diff
@@ -752,6 +727,208 @@ def update_bias_compressed_tensors(weight: torch.Tensor, scale: torch.Tensor, st
...
+        _, _, w2, w2_scale = self._get_mlp_weights(mlp_compute_input.layer)
+        bias1, bias2 = self._get_mlp_scale_bias(mlp_compute_input.layer)
+        group_list, group_list_type, output_dtype = self._maybe_convert_group_list(mlp_compute_input, bias1)
+        return torch_npu.npu_grouped_matmul(
+            x=[hidden_states],
+            weight=w2,
+            scale=w2_scale,
+            bias=bias2,
+            per_token_scale=[act_out_scale],
+            split_item=2,
```
Comment (zhangc10) [inline vllm_ascend/quantization/methods/w4a8/w4a8.py:544]: Can use `torch_npu.npu_grouped_matmul` directly.

File: vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py [hunk 15]
```diff
@@ -231,3 +237,85 @@ def process_weights_after_loading(self, layer):
...
+        hidden_states, dynamic_scale = torch_npu.npu_dynamic_mx_quant(
+            hidden_states, dst_type=torch_npu.float4_e2m1fn_x2
+        )
+        return hidden_states, maybe_normalize_mxfp_scale_layout(dynamic_scale)
+
+    def apply_gmm2(self, mlp_compute_input: MoEMlpComputeInput, hidden_states, act_out_scale):
+        layer = mlp_compute_input.layer
+        input_dtype = mlp_compute_input.hidden_states.dtype
+        use_bf16 = input_dtype in [torch.bfloat16, torch.uint8]
+        output_dtype = (
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py:289]: The `output_dtype` can be normalized using global variables, and the judgment conditions such as `use_bf16` are no longer required.

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py [hunk 16]
```diff
@@ -201,3 +203,95 @@ def process_weights_after_loading(self, layer):
...
+    def apply_gmm1_act_quant(self, mlp_compute_input: MoEMlpComputeInput):
+        hidden_states, pertoken_scale = self._quant_hidden_states(
+            mlp_compute_input.hidden_states, mlp_compute_input.dynamic_scale
+        )
+        layer = mlp_compute_input.layer
+        assert layer is not None
+        hidden_states = torch_npu.npu_grouped_matmul(
+            x=[hidden_states],
+            weight=[layer.w13_weight],
+            scale=None,
```
Comment (weijinqian0) [inline vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py:242]: The quantization method is not w4A8, so scaling should be required. Please check if there is an issue.

File: vllm_ascend/ops/fused_moe/dataclass/fused_experts.py [hunk 17]
```diff
@@ -68,10 +73,11 @@ def build_fused_experts_input(
...
     topk_weights: torch.Tensor,
     topk_ids: torch.Tensor,
-    w1: torch.Tensor | list[torch.Tensor],
-    w2: torch.Tensor | list[torch.Tensor],
+    layer=None,
     quant_type: QuantType,
     dynamic_eplb: bool,
+    w1: torch.Tensor | list[torch.Tensor] | None = None,
+    w2: torch.Tensor | list[torch.Tensor] | None = None,
     expert_map: torch.Tensor | None = None,
```
Comment (weijinqian0) [inline vllm_ascend/ops/fused_moe/dataclass/fused_experts.py:81]: Input parameters related to the weight need to be deleted.

**Review by weijinqian0 — DISMISSED** [review]

**Comment by earthmanylf** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by earthmanylf** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Comment by earthmanylf** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: still in progress, retry /rerun after completion

**Review by weijinqian0 — DISMISSED** [review]

**Comment by earthmanylf** [comment]
/rerun
[Bot]: rerun completed.

**Failed:**
- E2E: some jobs could not be re-run (still waiting on dependencies)

**Comment by earthmanylf** [comment]
/cancel
[Bot]: successfully force-cancelled the following runs:


https://github.com/vllm-project/vllm-ascend/actions/runs/32352728830

**Comment by earthmanylf** [comment]
/cancel
[Bot]: successfully force-cancelled the following runs:


https://github.com/vllm-project/vllm-ascend/actions/runs/32460976707

**Comment by earthmanylf** [comment]
/cancel
[Bot]: successfully force-cancelled the following runs:


https://github.com/vllm-project/vllm-ascend/actions/runs/33710941858

**Comment by earthmanylf** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by kunpengW-code — APPROVED** [review]
